From ea70959d61952545756c029e6771f5e31e7af8d6 Mon Sep 17 00:00:00 2001
From: Daniel Cederman <cederman@gaisler.com>
Date: Fri, 2 Jul 2021 11:17:26 +0200
Subject: [PATCH 7/7] serial: apbuart: Make it compatible with 8250

Signed-off-by: Daniel Cederman <cederman@gaisler.com>
---
 drivers/tty/serial/apbuart.c | 8 +++-----
 1 file changed, 3 insertions(+), 5 deletions(-)

diff --git a/drivers/tty/serial/apbuart.c b/drivers/tty/serial/apbuart.c
index a813008c1e7d..989f59917375 100644
--- a/drivers/tty/serial/apbuart.c
+++ b/drivers/tty/serial/apbuart.c
@@ -506,7 +506,7 @@ static int __init apbuart_console_setup(struct console *co, char *options)
 static struct uart_driver grlib_apbuart_driver;
 
 static struct console grlib_apbuart_console = {
-	.name = "ttyS",
+	.name = "ttyGR",
 	.write = apbuart_console_write,
 	.device = uart_console_device,
 	.setup = apbuart_console_setup,
@@ -530,10 +530,8 @@ console_initcall(apbuart_console_init);
 
 static struct uart_driver grlib_apbuart_driver = {
 	.owner = THIS_MODULE,
-	.driver_name = "serial",
-	.dev_name = "ttyS",
-	.major = SERIAL_APBUART_MAJOR,
-	.minor = SERIAL_APBUART_MINOR,
+	.driver_name = "gaisler-serial",
+	.dev_name = "ttyGR",
 	.nr = UART_NR,
 	.cons = APBUART_CONSOLE,
 };
-- 
2.25.1

